packages/task/src/browser/task-terminal-widget-manager.ts

```diff
@@ -0,0 +1,203 @@
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the Eclipse
+ * Public License v. 2.0 are satisfied: GNU General Public License, version 2
+ * with the GNU Classpath Exception which is available at
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ********************************************************************************/
+
+import { inject, injectable, postConstruct } from 'inversify';
+import { ApplicationShell, WidgetOpenerOptions } from '@theia/core/lib/browser';
+import { TerminalWidget } from '@theia/terminal/lib/browser/base/terminal-widget';
+import { TerminalWidgetFactoryOptions } from '@theia/terminal/lib/browser/terminal-widget-impl';
+import { TerminalService } from '@theia/terminal/lib/browser/base/terminal-service';
+import { PanelKind, TaskConfiguration, TaskWatcher, TaskExitedEvent, TaskServer } from '../common';
+import { TaskDefinitionRegistry } from './task-definition-registry';
+import { WorkspaceService } from '@theia/workspace/lib/browser/workspace-service';
+
+export interface TaskTerminalWidgetOpenerOptions extends WidgetOpenerOptions, TerminalWidgetFactoryOptions {
+    taskId: number;
+    taskPanel?: PanelKind;
+    taskConfig?: TaskConfiguration;
+}
+export namespace TaskTerminalWidgetOpenerOptions {
+    export function isDedicatedTerminal(options: TaskTerminalWidgetOpenerOptions): boolean {
+        return !!options.taskPanel && options.taskPanel === PanelKind.Dedicated;
+    }
+
+    export function isNewTerminal(options: TaskTerminalWidgetOpenerOptions): boolean {
+        return !!options.taskPanel && options.taskPanel === PanelKind.New;
+    }
+
+    export function isSharedTerminal(options: TaskTerminalWidgetOpenerOptions): boolean {
+        return options.taskPanel === undefined || options.taskPanel === PanelKind.Shared;
+    }
+}
+
+@injectable()
+export class TaskTerminalWidgetManager {
+
+    @inject(ApplicationShell)
+    protected readonly shell: ApplicationShell;
+
+    @inject(TaskDefinitionRegistry)
+    protected readonly taskDefinitionRegistry: TaskDefinitionRegistry;
+
+    @inject(TerminalService)
+    protected readonly terminalService: TerminalService;
+
+    @inject(TaskWatcher)
+    protected readonly taskWatcher: TaskWatcher;
+
+    @inject(TaskServer)
+    protected readonly taskServer: TaskServer;
+
+    @inject(WorkspaceService)
+    protected readonly workspaceService: WorkspaceService;
+
+    // Map indexed by terminal widget id
+    protected terminalWidgetMap: Map<string, { isDedicated: boolean, taskId: number, taskConfig?: TaskConfiguration, isIdle: boolean }> = new Map();
```

Comment:
It feels like properties of task terminal. I wonder why don't replace it with `TaskTerminalWidget` namespace, i.e.:
```ts
export interface TaskTerminalWidget extends TerminalWidget {
    readonly kind: 'task'
    dedicated?: boolean
    taskId?: boolean
    taskConfig?: TaskConfiguration
    busy?: boolean
}
export namespace TaskTerminalWidget {
    export function is(terminalWidget: TerminalWidget): TerminalWidget is TaskTerminalWidget {
        return terminalWidget.kind === ' task';
    }
}
```

Please also use `isDedicated` for method names, properties should be named `dedicated`. `idle` is fine, but in other places we use `busy`.

Reply:
the "terminal kind" can only be "user" and "other". If we want to have "user", "task", and more, we will need to store it in the local strage when browser refreshes.

Reply:
yes, It will be stored since it belongs to `TerminalWidgetOptions`.